### ISSUE TITLE
Feature request: Test `pip install`

### ISSUE THREAD

**Issue by cool-RR** [cool-RR/PySnooper#39]
@bittner , this is right up your alley so if you'll feel like doing it, it'll be cool.

It'll be nice if after every release, or after every push to master, there will be separate tests that do `pip install pysnooper` on all platforms and run the tests. This is to rule out any errors in the installation.

**Comment by bittner** [comment]
Every Python environment created by Tox gets `pysnooper` installed. The workflow is roughly:

1. Create virtualenv for Python x.y (e.g. Python 3.7 for `py37`)
1. Run `python setup.py install` (i.e. install `pysnooper`)
1. Install the dependencies (from `deps = ...` in the Tox environment)
1. Run `pytest` (i.e. from `commands = ...`)

If I'm not completely mistaken `pip install pysnooper` does nothing else than download the package from PyPI and run `python setup.py install`. Hence, in theory installing should succeed when Tox runs successfully. And this is now ensured.

Any doubt about this?

**Comment by bittner** [comment]
What is missing indeed is that tests are run on Windows and macOS. We've done this with PythonTurtle already, it's tedious but it's feasible, it's not magic.

And more tests, e.g. integration tests, are always a good idea. Sure!

**Comment by cool-RR** [comment]
Are you sure tox does `setup.py install`? Where is it written?

**Comment by bittner** [comment]
You may want to take a look at the Tox [System Overview](https://tox.readthedocs.io/en/latest/index.html#system-overview) and the related explanations in the Tox documentation.

**Comment by cool-RR** [comment]
Okay, thanks. I'd still welcome a direct test of `pip install pysnooper`. Low priority.

**Comment by enadol** [comment]
On Ubuntu/Raspbian, pip install pysnooper shoots the following:

pi@raspberrypi:~ $ pip3 install pysnooper
Downloading/unpacking pysnooper
  Downloading PySnooper-0.0.22-py2.py3-none-any.whl
Installing collected packages: pysnooper
Cleaning up...
Exception:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/basecommand.py", line 122, in main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/commands/install.py", line 295, in run
    requirement_set.install(install_options, global_options, root=options.root_path)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 1436, in install
    requirement.install(install_options, global_options, *args, **kwargs)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 672, in install
    self.move_wheel_files(self.source_dir, root=root)
  File "/usr/lib/python3/dist-packages/pip/req.py", line 902, in move_wheel_files
    pycompile=self.pycompile,
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 214, in move_wheel_files
    clobber(source, lib_dir, True)
  File "/usr/lib/python3/dist-packages/pip/wheel.py", line 204, in clobber
    os.makedirs(destdir)
  File "/usr/lib/python3.4/os.py", line 237, in makedirs
    mkdir(name, mode)
PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.4/dist-packages/pysnooper'

Storing debug log for failure in /home/pi/.pip/pip.log
----------------------------------
All this on a Rasbberry Pi

**Comment by cool-RR** [comment]
@enadol :

1. Please move this to a new issue.
2. Please confirm that you can install other packages via pip, and show the output

**Comment by enadol** [comment]
On a Windows machine with Anaconda, pip install pysnooper worked just fine:

![image](https://user-images.githubusercontent.com/5300289/56849808-d4550d80-68f9-11e9-89c4-8e011c3b395d.png)

I posted these messages in response to the request of testing pip install pysnooper. I wasn't trying to report the issue in Ubuntu per se. Sorry of it bothered.

**Comment by bittner** [comment]
@enadol: This behavior is normal on any GNU/Linux system.

> PermissionError: [Errno 13] Permission denied: '/usr/local/lib/python3.4/dist-packages/pysnooper'

The error says, you - user `pi` - are not allowed to write in areas that require elevated privileges (read: administrator access, i.e. root). Note that this is fine, it makes it more difficult for attackers that manage to enter your Raspberry Pi via an unprivileged account to do harm to your machine.

On Windows you may be used to work with an administrator-level user. This allows you to install programs anywhere, which makes your system insecure by default.

TL;DR - How install a Python package?
------

You have two options to install Python packages on GNU/Linux. As a super-user:

```console
$ sudo pip3 install pysnooper
```

... which will ask you for the `root` user's password, or in your user space:

```console
$ pip3 install --user pysnooper
```

The latter typically makes packages available in `~/.local/lib/` and scripts in `~/.local/bin/` (just in case this is not in your `PATH` by default).

**Comment by enadol** [comment]
Great! Thanks a lot! Cheers!

**Comment by awarecheetah** [comment]
Does anyone mind if I pick this one up?  It seems like a good first issue to tackle for new contributors.

**Comment by cool-RR** [comment]
It's a difficult one. But go ahead.

**Comment by awarecheetah** [comment]
With the existing automation that's already in place with tox and travis, I think the solution was easier to implement than originally thought--unless I'm missing something?
